# 2 задание

In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_boston
boston=load_boston()
X = pd.DataFrame(boston['data'], columns=boston['feature_names'])
y = pd.DataFrame(boston['target'], columns=['price'])

In [2]:
from sklearn.ensemble import RandomForestRegressor

In [6]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [7]:
model = RandomForestRegressor(n_estimators=1000, max_depth=12, random_state=42)


In [8]:
model.fit(X_train,y_train.values[:, 0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

In [9]:
y_pred=model.predict(X_test)

In [10]:
check_test = pd.DataFrame({"y_test": y_test["price"], "y_pred": y_pred.flatten(),})

check_test.head(10)

y_test     y_pred
173    23.6  22.806412
274    32.4  31.131464
491    13.6  16.339125
72     22.8  23.810726
452    16.1  17.139521
76     20.0  21.832284
316    17.8  19.895747
140    14.0  14.754118
471    19.6  21.240835
500    16.8  20.898658

In [11]:
from sklearn.metrics import r2_score
r2_score(y_test,y_pred)

0.87472606157312

R2=0.7112260057484973 в первой модели
R2=0.87472606157312 во второй модели.
Вторая модель работает лучше, так как его R2 ближе к 1

## 3 задание

In [14]:
help(RandomForestRegressor.feature_importances_)

Help on property:

    The impurity-based feature importances.
    
    The higher, the more important the feature.
    The importance of a feature is computed as the (normalized)
    total reduction of the criterion brought by that feature.  It is also
    known as the Gini importance.
    
    high cardinality features (many unique values). See
    :func:`sklearn.inspection.permutation_importance` as an alternative.
    
    Returns
    -------
    feature_importances_ : ndarray of shape (n_features,)
        The values of this array sum to 1, unless all trees are single node
        trees consisting of only the root node, in which case it will be an
        array of zeros.



In [16]:
feature_importance = pd.DataFrame({'name':X.columns, 
                                   'feature_importance':model.feature_importances_})
feature_importance

name  feature_importance
0      CRIM            0.031676
1        ZN            0.001543
2     INDUS            0.007138
3      CHAS            0.001236
4       NOX            0.014269
5        RM            0.402682
6       AGE            0.014299
7       DIS            0.063973
8       RAD            0.005281
9       TAX            0.011525
10  PTRATIO            0.018081
11        B            0.012451
12    LSTAT            0.415847

In [17]:
feature_importance.nlargest(2, 'feature_importance')

name  feature_importance
12  LSTAT            0.415847
5      RM            0.402682